src/app/reporting/Engine.cpp

```diff
@@ -100,15 +108,38 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeDataList(ReportData::Builder &
     {
         if (clusterInfo->IsDirty())
         {
-            AttributeDataElement::Builder attributeDataElementBuilder = attributeDataList.CreateAttributeDataElementBuilder();
             if (apReadHandler->IsInitialReport())
             {
                 // Retrieve data for this cluster instance and clear its dirty flag.
-                err = RetrieveClusterData(attributeDataElementBuilder, *clusterInfo);
+                err = RetrieveClusterData(attributeDataList, *clusterInfo);
                 VerifyOrExit(err == CHIP_NO_ERROR,
                              ChipLogError(DataManagement, "<RE:Run> Error retrieving data from cluster, aborting"));
                 attributeClean = false;
             }
+            else
+            {
+                for (auto & path : mDirtyPathList)
+                {
+                    if (path.IsDirty())
+                    {
+                        if (clusterInfo->IsAttributePathSupersetOf(path))
+                        {
+                            err = RetrieveClusterData(attributeDataList, path);
+                        }
+                        else if (path.IsAttributePathSupersetOf(*clusterInfo))
+                        {
+                            err = RetrieveClusterData(attributeDataList, *clusterInfo);
```

Comment:
does this mean that if one element changes, we report all elements? I.e. if I ask for EP 1, CL 2 all attributes, then if a single attribute changes I get all attribute values? Should this not be filtered by path in some way?

Reply:
1. if someone mutate one element(path)in server and that element is part of interested cluster path(clusterInfo), then we just report that particular element change, this code is "clusterInfo->IsAttributePathSupersetOf(path)", 

2. if someone mutate one root change(path), which is super set of small interested cluster path, then we only report that small interested cluster path, this code is "path.IsAttributePathSupersetOf(*clusterInfo“